tests/unit/test_scrape_jobs.py

```diff
@@ -0,0 +1,81 @@
+from charm import AlertmanagerCharm
+from helpers import k8s_resource_multipatch
+from ops.testing import Harness
+
+
+class TestWithInitialHooks(unittest.TestCase):
+    container_name: str = "alertmanager"
+
+    @patch("lightkube.core.client.GenericSyncClient")
+    @patch.object(WorkloadManager, "check_config", lambda *a, **kw: ("ok", ""))
+    @k8s_resource_multipatch
+    @patch.object(WorkloadManager, "_alertmanager_version", property(lambda *_: "0.0.0"))
+    def setUp(self, *unused):
+        self.harness = Harness(AlertmanagerCharm)
+        self.addCleanup(self.harness.cleanup)
+
+        self.harness.set_leader(True)
+        self.app_name = "alertmanager-k8s"
+        # Create the peer relation before running harness.begin_with_initial_hooks(), because
+        # otherwise it will create it for you and we don't know the rel_id
+        self.peer_rel_id = self.harness.add_relation("replicas", self.app_name)
+
+        self.harness.begin_with_initial_hooks()
+
+    @patch.object(AlertmanagerCharm, "_internal_url", new_callable=PropertyMock)
+    @patch.object(AlertmanagerCharm, "_scheme", new_callable=PropertyMock)
+    def test_self_scraping_job_with_no_peers(self, _mock_scheme, _mock_internal_url):
+        scheme = "https"
+        _mock_scheme.return_value = scheme
+        url_no_scheme = f"test-internal.url:{self.harness.charm._ports.api}"
+        _mock_internal_url.return_value = f"{scheme}://{url_no_scheme}"
+        jobs_expected = [
+            {
+                "metrics_path": "/metrics",
+                "scheme": scheme,
+                "static_configs": [{"targets": [url_no_scheme]}],
+            }
+        ]
+
+        jobs = self.harness.charm.self_scraping_job
+        self.assertEqual(jobs, jobs_expected)
+
+    @patch.object(AlertmanagerCharm, "_internal_url", new_callable=PropertyMock)
+    @patch.object(AlertmanagerCharm, "_scheme", new_callable=PropertyMock)
+    def test_self_scraping_job_with_peers(self, _mock_scheme, _mock_internal_url):
+        scheme = "https"
+        _mock_scheme.return_value = scheme
+
+        targets = [
+            f"test-internal-0.url:{self.harness.charm._ports.api}",
+            f"test-internal-1.url:{self.harness.charm._ports.api}",
+            f"test-internal-2.url:{self.harness.charm._ports.api}",
+        ]
+        metrics_path = "/metrics"
+        _mock_internal_url.return_value = f"{scheme}://{targets[0]}"
+
+        jobs_expected = [
+            {
+                "metrics_path": metrics_path,
+                "scheme": scheme,
+                "static_configs": [{"targets": targets}],
+            }
+        ]
+
+        # Add peers
+        for i, target in enumerate(targets[1:], 1):
+            unit_name = f"{self.app_name}/{i}"
+            self.harness.add_relation_unit(self.peer_rel_id, unit_name)
+            self.harness.update_relation_data(
+                self.peer_rel_id, unit_name, {"private_address": f"{scheme}://{target}"}
+            )
```

Comment:
We only add two units. I think it's better for clarity to be more explicit.

```suggestion
        self.harness.add_relation_unit(self.peer_rel_id, "am/1")
        self.harness.update_relation_data(self.peer_rel_id, "am/1", {"private_address": f"https://am-1.demo.internal:9093"})
        self.harness.add_relation_unit(self.peer_rel_id, "am/2")
        self.harness.update_relation_data(self.peer_rel_id, "am/2", {"private_address": f"https://am-2.demo.internal:9093"})
```
Same line count, but better clarity.

Reply:
Something I'd like to avoid is having string literals that rely each other.  For instance, how `"am/1"` is written as a literal in a few places, and how the `targets` passed into the mock are written as separate literals from the `jobs_expected` literals.  Do you have a suggestion that feels better to you but also avoids this?

I took a second attempt at this, but I think it got worse... 
```python
    @patch.object(AlertmanagerCharm, "_internal_url", new_callable=PropertyMock)
    @patch.object(AlertmanagerCharm, "_scheme", new_callable=PropertyMock)
    def test_self_scraping_job_with_peers(self, _mock_scheme, _mock_internal_url):
        scheme = "https"
        _mock_scheme.return_value = scheme

        this_unit_internal_url = f"{scheme}://test-internal-0.url:1234"
        _mock_internal_url.return_value = this_unit_internal_url

        peer_unit_addresses = {
            "am/1": f"{scheme}://test-internal-1.url:1234",
            "am/2": f"{scheme}://test-internal-2.url:1234",
        }

        targets = [this_unit_internal_url] + [
            address.replace(f"{scheme}://", "") for address in peer_unit_addresses.values()
        ]

        jobs_expected = [
            {
                "metrics_path": "/metrics",
                "scheme": scheme,
                "static_configs": [{"targets": targets}],
            }
        ]

        # This could also be broken out like your first suggestion.  I think this is clearer, but happy to switch
        for unit_name, address in peer_unit_addresses.items():
            self.harness.add_relation_unit(self.peer_rel_id, unit_name)
            self.harness.update_relation_data(
                self.peer_rel_id, unit_name, {"private_address": address}
            )

        jobs_actual = self.harness.charm.self_scraping_job
        self.assertEqual(jobs_expected, jobs_actual)
```